shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala

```diff
@@ -231,27 +291,126 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
         new ColumnarBatch(cols.map(_.copyToHost()).toArray, gpuBatch.numRows())
       }
     })
-    cbRdd.mapPartitions(iter => new CloseableColumnBatchIterator(iter))
+    cbRdd.mapPartitions(iter => CloseableColumnBatchIterator(iter))
   }
 
   /**
    * Convert the cached batch into `InternalRow`s.
+   *
    * @param input the cached batches that should be converted.
    * @param cacheAttributes the attributes of the data in the batch.
    * @param selectedAttributes the field that should be loaded from the data and the order they
    *                           should appear in the output rows.
    * @param conf the configuration for the job.
    * @return RDD of the rows that were stored in the cached batches.
    */
-  override def convertCachedBatchToInternalRow(input: RDD[CachedBatch],
-                                cacheAttributes: Seq[Attribute],
-                                selectedAttributes: Seq[Attribute],
-                                conf: SQLConf): RDD[InternalRow] = {
-    val cb = convertCachedBatchToColumnarBatch(input, cacheAttributes, selectedAttributes, conf)
-    val rowRdd = cb.mapPartitions(iter => {
-      new ColumnarToRowIterator(iter)
-    })
-    rowRdd
+  override def convertCachedBatchToInternalRow(
+     input: RDD[CachedBatch],
+     cacheAttributes: Seq[Attribute],
+     selectedAttributes: Seq[Attribute],
+     conf: SQLConf): RDD[InternalRow] = {
+
+    val sparkSession = SparkSession.active
+    val hadoopConf = getHadoopConf(selectedAttributes.toStructType, conf)
+    val broadcastedHadoopConf =
+      sparkSession.sparkContext.broadcast(new SerializableConfiguration(hadoopConf))
+
+    input.mapPartitions {
+      cbIter => {
+        new Iterator[InternalRow]() {
+          var iter: Iterator[InternalRow] = null
+
+          override def hasNext: Boolean = {
+            // either the current iterator has an element,
+            // else go over the batch and get the next non-degenerate iterator
+            (iter != null && iter.hasNext) ||
+            {
+              while (cbIter.hasNext && !iter.hasNext) {
+                iter = convertColumnarBatchToInternalRowIter()
+              }
+              iter.hasNext
+            }
+          }
+
+          override def next(): InternalRow = {
+            // will return the next InternalRow if hasNext() is true, otherwise undefined behavior
+            iter.next()
```

Comment:
Could we NPE here if cbIter.hasNext is null from the very beginning and we were never able to instantiate a valid `iter`?  (Thinking of caching the degenerate table case of zero rows, if that is possible.)

Reply:
yes it will if we don't call `hasNext` before calling next. 

I don't know what is the right thing to do here. Should we call `doesNextValidIteratorHasNext` method? We could still theoretically get into a case where even after calling that method we don't have a valid iterator. What do we do in that case?

Reply:
> What do we do in that case?

Throw `NoSuchElementException` as most other iterators do when calling `next()` beyond the end of the iterator.